File: Java/Recursion-1/stringClean.java

public class recursion1 {
	
	// given a string, return recursively a "cleaned" string where adjacent chars that are the same have been reduced to a single char
	// so "yyzzza" yields "yza"
	public String stringClean(String str) {
		
		if (str.length() == 0) {
			return "";
		}
		
		if (str.length() > 1 && str.charAt(0) == str.charAt(1)) {
			return stringClean(str.substring(1));
		}
		else {
			return str.charAt(0) + stringClean(str.substring(1));
		}
		
	}

	public static void main(String[] args) {

		recursion1 o = new recursion1();
		
		System.out.println(o.stringClean("yyzzza"));
		System.out.println(o.stringClean("abbbcdd"));
		System.out.println(o.stringClean("Hello"));
		
	}

}

File: Python/Recursion-1/string_clean.py

# given a string, return recursively a "cleaned" string where adjacent chars that are the same have been reduced to a single char
# so "yyzzza" yields "yza"
def string_clean(str):
	if not str:
		return ''

	if len(str) > 1 and str[0] == str[1]:
		return string_clean(str[1:])
	else:
		return str[0] + string_clean(str[1:])

print(string_clean('yyzzza'))
print(string_clean('abbbcdd'))
print(string_clean('Hello'))
